paddle/fluid/imperative/layer.cc

```diff
@@ -240,189 +350,452 @@ std::unique_ptr<VarBase> VarBase::NewVarBase(const platform::Place& dst_place,
-            VLOG(2) << "insert new map for " << origin_outputs[i]->Name();
-            std::pair<platform::Place, std::vector<std::pair<int, VarBase*>>>
-                tmp(place_, {std::make_pair(this->trace_id_, outputs[i])});
-            bck_map->insert(std::make_pair(origin_outputs[i], tmp));
-          }
+    if (ref_cnt_ == 1) {
+      *(var_->MutableVar()) = std::move(*(var->MutableVar()));
+    } else {
+      if (tmp_grad_vars_.empty()) {
+        tmp_grad_vars_.reserve(ref_cnt_);
+      }
 
-          PADDLE_ENFORCE(grad_ref->find(origin_outputs[i]) != grad_ref->end(),
-                         "Can't find  %s in grad_reference count map",
-                         origin_outputs[i]->Name());
-          PADDLE_ENFORCE(grad_ref->at(origin_outputs[i]) >= 1,
-                         "Backward error when calculate grad reference");
-          if (grad_ref->at(origin_outputs[i]) > 1) {
-            VLOG(2) << "remove ref for " << origin_outputs[i]->Name();
-            grad_ref->at(origin_outputs[i])--;
-          } else {
-            VLOG(2) << "Add grad for: " << origin_outputs[i]->Name();
-            AddGradBySort(bck_map, origin_outputs[i]);
-            grad_ref->at(origin_outputs[i])--;
-          }
-        } else {
-          framework::Variable* grad = outputs[i]->var_.get();
-          framework::Variable* orig_grad = origin_outputs[i]->var_.get();
-          VLOG(2) << "AddTo Called with orig_grad is: "
-                  << origin_outputs[i]->name_ << " Grad to be added is "
-                  << outputs[i]->name_;
-          AddTo(grad, orig_grad, place_);
-          delete outputs[i];
-        }
+      tmp_grad_vars_.emplace_back(std::move(var), trace_id);
+      if (tmp_grad_vars_.size() != ref_cnt_) {
+        return;
+      }
+
+      std::sort(tmp_grad_vars_.begin(), tmp_grad_vars_.end(),
+                [](const std::pair<std::shared_ptr<VarBase>, size_t>& p1,
+                   const std::pair<std::shared_ptr<VarBase>, size_t>& p2) {
+                  return p1.second > p2.second;
+                });
+
+      *(var_->MutableVar()) =
+          std::move(*(tmp_grad_vars_[0].first->MutableVar()));
+      for (size_t i = 1; i < tmp_grad_vars_.size(); ++i) {
+        TensorAdd(tmp_grad_vars_[i].first->Var(), var_->MutableVar());
       }
+
+      tmp_grad_vars_.clear();
     }
   }
 
-  return input_vars_;
+  size_t IncreaseRefCnt() { return ++ref_cnt_; }
+
+ private:
+  VarBase* var_;
+  detail::BackwardStrategy backward_strategy_;
+  size_t ref_cnt_{0};
+  std::vector<std::pair<std::shared_ptr<VarBase>, size_t>> tmp_grad_vars_;
+  std::vector<size_t> trace_ids_;
+};
+
+class AutoGradImpl {
+ public:
+  explicit AutoGradImpl(OpBase* op, const detail::BackwardStrategy& strategy);
+
+  void operator()();
+
+ private:
+  void PrepareDeps();
+
+  bool CheckBackwardInputs(OpBase* op);
+
+  void CheckBackwardOutputs(OpBase* op);
+
+  void SumGradient(OpBase* op, std::shared_ptr<VarBase> src, VarBase* dst) {
```

Comment:
why `share_ptr` here?

Reply:
Because `Tracer` needs the input type to be `std::shared_ptr`.